### PR TITLE
Add groups in User model

### PR DESCRIPTION
**Pull request by cychu42**
This is for database part of issue #290
This is meant to have `groups` as a column to store role(s)/group(s) they have, as well as related changes.

## Changes:
* Add groups in `User` model as String in `prisma\schema.prisma`
* Add examples in `prisma\seed.ts`
* Modify `createUser()` in `app\models\user.server.ts` accordingly
* Add `isStudent()`, `isFaculty()`, and `isAdmin()` in `app\models\user.server.ts`.
* Combine `firstName` and `lastName` into `displayName`
* Modify arguments of `createUser()` in `app\routes\login\callback.tsx`

`isStudent()`, `isFaculty()`, and `isAdmin()` accept `username` as parameter. They would return `true` if they have the role/group, `false` if not.

## Testing:
You make a local copy of the project branch. If you follow the contribution document's steps to set it up, you can then check whether the schema and seeding work, via `npm run db:studio`.
You can use the functions `isStudent()`, `isFaculty()`, and `isAdmin()` to test them in your code, or test in a [query file](https://github.com/DevelopingSpace/starchart/wiki/Prisma#3-write-your-first-query).

### CODE REVIEW

File: app/models/user.server.ts [hunk 1]
```diff
@@ -52,7 +52,8 @@ export async function createUser(
   username: PrismaUser['username'],
   firstName: PrismaUser['firstName'],
```
Comment (humphd) [inline app/models/user.server.ts:53]: Drop `firstName` and `lastName` and use `displayName` or just `name` if you want (both parts are given to us in 1 single name).

File: app/models/user.server.ts [hunk 2]
```diff
@@ -52,7 +52,8 @@ export async function createUser(
...
   lastName: PrismaUser['lastName'],
-  email: PrismaUser['email']
+  email: PrismaUser['email'],
+  groups: PrismaUser['groups']
```
Comment (humphd) [inline app/models/user.server.ts:56]: Change this to `group` (singular)

File: app/models/user.server.ts [hunk 3]
```diff
@@ -61,6 +62,51 @@ export async function createUser(
...
       lastName,
       email,
+      groups,
     },
   });
 }
+
+export async function isStudent(username: PrismaUser['username']) {
+  let result;
+  prisma.user.findUnique({ where: { username } }).then((user) => {
```
Comment (humphd) [inline app/models/user.server.ts:72]:
```ts
const { group } = await prisma.user.findUniqueOrThrow({ where: { username } });
// The group will have -dev in it on staging but not on prod
return /mycustomdomain(-dev)?-students/.test(group);
```

File: app/models/user.server.ts [hunk 4]
```diff
@@ -61,6 +62,51 @@ export async function createUser(
...
+      result = true;
+    } else {
+      result = false;
+    }
+  });
+  return result;
+}
+
+export async function isFaculty(username: PrismaUser['username']) {
+  let result;
```
Comment (humphd) [inline app/models/user.server.ts:85]: See pattern I wrote above...

File: app/models/user.server.ts [hunk 5]
```diff
@@ -61,6 +62,51 @@ export async function createUser(
...
+      result = true;
+    } else {
+      result = false;
+    }
+  });
+
+  return result;
+}
+
+export async function isAdmin(username: PrismaUser['username']) {
```
Comment (humphd) [inline app/models/user.server.ts:99]: See pattern I wrote above...

File: app/models/user.server.ts [hunk 6]
```diff
@@ -61,6 +62,51 @@ export async function createUser(
...
+  let result;
+  prisma.user.findUnique({ where: { username } }).then((user) => {
+    if (!user) {
+      // Leave it undefined
+    } else if (user.groups.includes('mycustomdomain-Admins')) {
```
Comment (humphd) [inline app/models/user.server.ts:104]: lower case 'a' in admins

File: prisma/seed.ts [hunk 7]
```diff
@@ -26,9 +27,29 @@ async function seed() {
...
+        lastName: 'Galilei',
+        email: 'user2@myseneca.ca',
+        groups: 'mycustomdomain-students',
+      },
+      {
+        username: 'user3',
+        firstName: 'Adam',
+        lastName: 'Mason',
+        email: 'user3@myseneca.ca',
+        groups: 'mycustomdomain-faculty, mycustomdomain-admins',
```
Comment (humphd) [inline prisma/seed.ts:51]: It will only ever be a single string

**Review by humphd — CHANGES_REQUESTED** [review]
See my comment in https://github.com/DevelopingSpace/starchart/issues/290#issuecomment-1460882772

**Review by SerpentBytes — CHANGES_REQUESTED** [review]
Type-checks seem to fail because fewer than expected arguments are being passed to a function call in `app/routes/login/callback.tsx`. Please take a look at the failing tests for more details.

File: app/models/user.server.ts (hunk 3 above)
Comment (cychu42) [reply app/models/user.server.ts:72]: Oh, I see what you meant by staging vs prod in that issue.

File: prisma/seed.ts (hunk 7 above)
Comment (cychu42) [reply prisma/seed.ts:51]: Are you saying an user will only have one group?

Comment (sfrunza13) [reply prisma/seed.ts:51]: This is also what I understood from our conversations yesterday. I think users will only ever have 1 group they belong to.

**Comment by cychu42** [comment]
This piece of code is failing in typcheck, due to changes to `createUser()`'s parameter for this PR.
```
   if (!(await checkUsernameExists(username))) {
    await createUser(
      username,
      samlResponse.attributes['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'],
      samlResponse.attributes['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'],
      samlResponse.attributes['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress']
    );
  }
```

File: app/models/user.server.ts (hunk 1 above)
Comment (cychu42) [reply app/models/user.server.ts]: Got it!

File: app/models/user.server.ts (hunk 6 above)
Comment (cychu42) [reply app/models/user.server.ts]: Ah, good catch.

**Comment by cychu42** [comment]
I modified those arguments to what Stefan will be using after he implements changes to IDP side.

**Review by SerpentBytes — DISMISSED** [review]

**Review by sfrunza13 — APPROVED** [review]

File: prisma/schema.prisma [hunk 8]
```diff
@@ -8,11 +8,10 @@ generator client {
...
 model User {
-  username  String @id
-  firstName String
-  lastName  String
-  email     String @unique
-
+  username    String       @id
+  displayName String
+  email       String       @unique
+  group      String
```
Comment (humphd) [inline prisma/schema.prisma:14]: Please from `prisma format` to clean up this formatting.

File: prisma/seed.ts [hunk 9]
```diff
@@ -23,12 +23,28 @@ async function seed() {
...
+      },
+      {
+        username: 'user2',
+        displayName: 'Galileo Galilei',
+        email: 'user2@myseneca.ca',
+        group: 'mycustomdomain-students',
+      },
+      {
+        username: 'user3',
+        displayName: 'Adam Mason',
```
Comment (humphd) [inline prisma/seed.ts:44]: cc @sfrunza13 - we better keep this seed list and the config for simplesaml and our docs in sync.

**Review by humphd — CHANGES_REQUESTED** [review]
Nice work.  One nit.

**Comment by sfrunza13** [comment]
Once this goes in I will pull it and rebase my feature branch then make my PR right after this.

File: prisma/schema.prisma (hunk 8 above)
Comment (cychu42) [reply prisma/schema.prisma]: Thanks. Done!

**Comment by cychu42** [comment]
OK. I formatted the schema file.

**Review by sfrunza13 — APPROVED** [review]

**Review by SerpentBytes — APPROVED** [review]

File: prisma/seed.ts (hunk 9 above)
Comment (humphd) [reply prisma/seed.ts:44]: Fixed in #312

**Review by humphd — APPROVED** [review]
👍

**Comment by cychu42** [comment]
Thank you all for the reviews.

**Pull request merged**